csrc/sycl/kernels.cpp

```diff
@@ -0,0 +1,5528 @@
+//{
+//// element-wise kernel
+//// 1. Load batch x k into registers
+//// 2. Load k x k into registers
+//// 3. dequantize and store in second pair of k x k
+//// 4. matmul
+//// 5. sum with cub
+//// 6. store outputs
+//// TC kernel
+//// use k warps per thread block
+//// 1. threadblock use read-only cache to read in register tile for A into shared memory
+//// 2. each warp loops over shared memory tiles of A of size 8x16 and loads them into fragments
+//// 3. each warp reads a segment of values 16x32 from B
+//// 4. do dequantization from register of B into second pair of registers
+//// 5. store (4) into fragment
+//// 6. matmul aggregate into fragment C
+//// 7. aggregate files of C into shared memory block C
+//// 8. sum (7)
+//// 9. write outputs to matmul output matrix
+//}
+
+template <typename T, typename TCAST, int ITEMS> inline void vector_load(T *local, T * __restrict__ const buffer, int idx, int limit_base, int limit, float zero_value = 0.0f)
+{
+    if(limit_base + ITEMS <= limit)
+      reinterpret_cast<TCAST*>(local)[0] = reinterpret_cast<TCAST*>(buffer)[idx/ITEMS];
+    else
+    {
+      for(int k = 0; k < ITEMS; k++)
+      {
+        if(limit_base + k < limit)
+          local[k] = buffer[idx+k];
+        else
+          local[k] = (T)zero_value;
+      }
+    }
+}
+
+#define WARPS 3
+/*
+DPCT1110:15: The total declared local variable size in device function gemm_device exceeds 128 bytes and may cause high register pressure. Consult with your hardware vendor to find the total register size available and adjust the code, or use smaller sub-group size to avoid high register pressure.
+*/
+template <typename T, int BITS, int THREADS> SYCL_EXTERNAL void gemm_device(int M, int N, int K, T * __restrict__ const A,  T* B,  T * out,  int lda, int ldb, int ldc, const sycl::nd_item<3> &item_ct1, T *smem_A, T *smem_B)
+{
+
+#if DPCT_COMPATIBILITY_TEMP >= 750
+	
+  int col_offset = item_ct1.get_group(2) *32;
+  const int warp_id = item_ct1.get_local_id(2) / 32;
+  const int half_warp_id = item_ct1.get_local_id(2) / 16;
+  const int half_warp_lane = item_ct1.get_local_id(2) % 16;
+  const int batch_size_warps = (WARPS-1)*2;
+  const int val_per_iter = item_ct1.get_local_range(2)-32;
+
+  T local_A[4];
+  T local_B[128];
+
+  const int a_tile_offset = 16;
+  const int b_tile_offset = (16*32 + 16);
+
+  
+  
+  //__shared__ T smem_C[8*32];
+
+   /*
+   DPCT1082:16: Migration of nvcuda::wmma::fragment<wmma::matrix_a, 8, 32, 16, half, wmma::row_major> type is not supported.
+   */
+   /*
+   DPCT1082:17: Migration of nvcuda::wmma::matrix_a type is not supported.
+   */
+   /*
+   DPCT1082:18: Migration of nvcuda::wmma::row_major type is not supported.
+   */
+   sycl::ext::oneapi::experimental::matrix::joint_matrix<sycl::sub_group, sycl::half, sycl::ext::oneapi::experimental::matrix::use::a, 8, 16, sycl::ext::oneapi::experimental::matrix::layout::row_major> a_frag;
```

Comment:
This code is only portable across different Nvidia devices: see https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc#nvidia-tensor-cores-supported-combinations

These parameter combinations (the matrix sizes) are only supported in the cuda backend. Unfortunately different matrix core vendors do not support the same range of sizes on matrix tiles. Hence, whilst the interfaces are portable, the parameters are not. So if you want to support more than cuda you need to write a new version for each additional backend (intel AMX,XMX or amd Matrix Cores) if you want to still use joint_matrix.

Reply:
For AMX or XMX, and other cores, does a runtime macro design pattern look ok?  I am trying to figure out how to link separate instruction calls for different vendors. Suggestions appreciated.

Reply:
The only other option I know of is to pass a backend query to the kernel that can be queried at runtime. MACRO might make more sense I guess for something like this. 
If you don't want to rely on a bitsandbytes macro then these backend macros get defined: depending on the build: https://github.com/codeplaysoftware/SYCL-samples/blob/main/src/MPI_with_SYCL/send_recv_buff.cpp#L16

Now in theory two could be defined at the same time.. but this might not be a concern.

Maybe there is some other option. But I don't know it.